### PR TITLE
Take path-like filepaths more broadly for datasets

### PR DESCRIPTION
**Pull request by deepyaman**
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description

Quick, backwards-compatible fix to resolve an inconsistency. E.g. `dask.ParquetDataset` may not fail when initialized with a `Path` file path, but `matplotlib.MatplotlibWriter` will, because of where `get_protocol_for_filepath` is called (buried in `exists` method, in the first case).

## Development notes

This doesn't claim to make `Path` objects be accepted as filepath everywhere; it's a quick fix to unblock `matplotlib` dataset doctest.

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass. Tested locally, and it works.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

File: kedro/io/core.py [hunk 1]
```diff
@@ -23,6 +23,9 @@
 from kedro import KedroDeprecationWarning
 from kedro.utils import load_obj
 
+if TYPE_CHECKING:
```
Comment (astrojuanlu) [inline kedro/io/core.py:26]: `os` is innocuous enough that maybe we don't need this conditional at all?

**Review by astrojuanlu — COMMENTED** [review]
Left one nitpick

**Review by merelcht — APPROVED** [review]
LGTM! 👍 I would personally add this to the release notes.

File: kedro/io/core.py (hunk 1 above)
Comment (deepyaman) [reply kedro/io/core.py]: @astrojuanlu Done!

**Comment by deepyaman** [comment]
> LGTM! 👍 I would personally add this to the release notes.

@merelcht Done!

**Comment by astrojuanlu** [comment]
```

**Comment by deepyaman** [comment]
> sphinx.errors.SphinxWarning: /home/docs/checkouts/readthedocs.org/user_builds/kedro/checkouts/3345/docs/source/kedro_project_setup/starters.md:147:broken link: https://github.com/kedro-org/kedro-starters/tree/main/pandas-iris/%7B%7B%20cookiecutter.repo_name%20%7D%7D/src/%7B%7B%20cookiecutter.python_package%20%7D%7D (404 Client Error: Not Found for url: https://github.com/kedro-org/kedro-starters/tree/main/pandas-iris/%7B%7B%20cookiecutter.repo_name%20%7D%7D/src/%7B%7B%20cookiecutter.python_package%20%7D%7D)

@astrojuanlu I'll repoint to `develop`? Do you know off the top of your head, does Kedro-Datasets CI work against `develop` anyway?

**Comment by astrojuanlu** [comment]
> Do you know off the top of your head, does Kedro-Datasets CI work against develop anyway? I want to say yes, but I haven't actually looked.

I don't remember if we have checked already, but it will be a necessary check before datasets 2.0.0 is released

**Review by astrojuanlu — APPROVED** [review]

**Pull request merged**